DOC: Trying to make a single place for contributor docs

mgeier opened the pull request:
See #1193.

willingc commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,162 @@
+
+Python Development Libraries
+""""""""""""""""""""""""""""
+
```
We should probably add a sentence at the start that mentions: For all operating systems, Python libraries can be downloaded from python.org https://www.python.org/downloads/

mgeier replied:
Should we actually lead beginners there?

I assume that if they have Python installed already (which, if anywhere, should be documented in the README and not in CONTRIBUTING), your link doesn't help them at all, right?

I can only speak about Linux, where this link definitely doesn't help any beginner.
Is this different for OSX or Windows?

mgeier replied:
I just tried it on a virtual Windows 10 installation with Anaconda.

I didn't get `bower` to work, but PyZMQ and Tornado were successfully installed without needing to install a compiler or any Python development libraries.

willingc commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,162 @@
+
+Or you can use your system's package manager ...
+
+If you use homebrew on Mac OS X::
```
Homebrew

mgeier replied:
Thanks, I'll actually add a link.

willingc commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,162 @@
+
+    brew install node
+
+For Debian/Ubuntu systems, you need to use the ``nodejs-legacy`` package and
```
~~need to~~ should  ... package ~~and not~~ instead of

mgeier replied:
OK

willingc commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,162 @@
+The development libraries might be needed for the installation of *PyZMQ*,
+*Tornado* and *Jinja2*.
+
+Alternatively -- if you prefer -- you can also install those packages directly
```
Using `pip`, as described in the beginning steps, is the **recommended** method of installing `PyZMQ`, `Tornado`, and `Jinja2` on your system.  Alternatively, Debian/Ubuntu users may install those ...

mgeier replied:
I actually tried this (see https://github.com/jupyter/notebook/issues/1193#issuecomment-196798100). I'll come back to that later ...

willingc commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,162 @@
+    cd notebook
+    pip install -e . --user
+
+If you want this to be available for all users of your system (assuming you
```
Replace `this` with `the development environment`

mgeier replied:
OK.

I just tried it in a Python 3 virtualenv and it seems that the `--user` option also works there, so I'll remove the comment about the virtualenv here.

willingc commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,162 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+There is a build step for the JavaScript and CSS in the notebook.
+You will need to run this command whenever there are changes to JavaScript or
```
To make sure that you are working with up-to-date code, you ...

mgeier replied:
OK. I hope I'll get this right. Please double-check after my next commit.

willingc commented on docs/source/development_release.rst:
```diff
@@ -8,11 +8,13 @@ notebook.
 ---------------------
-Review :ref:`development_js`. Make sure all the tools needed to generate the
+
+Review ``CONTRIBUTING.rst``. Make sure all the tools needed to generate the
```
Let's make sure that this is a clickable link.

mgeier replied:
I don't know how, that's why I showed only the file name.

The Sphinx-way to do that would probably be the `:download:` role, but that links to a *copy* of the file, which is not the file where potential changes are supposed to be made.

willingc commented on setup.py:
```diff
@@ -155,9 +155,14 @@
     ':sys_platform != "win32"': ['terminado>=0.3.3'],
-    'doc': ['Sphinx>=1.1'],
+    'doc': [
+        'Sphinx>=1.1',
```
Sphinx should really be 1.3.6 or above to properly support Markdown and RST as sources and avoid some errors found on kernel gateway project.

mgeier replied:
Is this already available on readthedocs?
Last time I checked it was at 1.3.5 ...

willingc replied:
We pip install 1.3.6 within our conda environment.yml for building the docs on RTD. This version is needed for markdown sources if one wishes to have the "Edit on GitHub" button in the right hand of the RTD theme direct to the correct place.

willingc replied:
Let's remove setup.py from this PR. We would like to keep this unchanged.

mgeier replied:
But we need the added dependencies for the shown instructions to work!

namely:

```
pip install -e .[test] --user
pip install -e .[doc] --user
```

I can make a separate PR if you prefer, but those changes are directly related to the documentation changes I made.

willingc replied:
For now, I would like to leave the setup.py file untouched and split it out from this PR. I realize that impacts the parts you mention. In the future, as part of a release, we can revisit the setup.py. 

As for the directions, you can comment them out and replace with a simple list of dependencies for the user to install when building docs or running tests or I can merge an updated PR with setup.py removed from the PR and iterate on the document. Either is cool with me.

Thanks! :sunrise:

mgeier replied:
The natural place for the dependencies is `setup.py`. Some of them were already there, I just added the missing ones.

Adding a list of dependencies to `CONTRIBUTING.rst` is a bad idea IMHO.
There should be only a single place for this list, and `setup.py` is the best candidate.
If you want me to add the list to `CONTRIBUTING.rst`, I'll have to remove *all* dependencies from `setup.py`, so either way, `setup.py` will have to be changed.

If you want, I can make a separate PR to add the dependencies to `setup.py`, but it'll have to be merged before this PR.

willingc commented:
@mgeier Overall, I think this is a nice improvement to the development installation docs.

I've made a few suggestions. I suspect @takluyver and @captainsafia will have comments as well. I would appreciate it if @jdfreder + @Carreau would look at the JS install and @minrk to look at the `setup.py` file. 

Thanks!

mgeier commented:
@willingc Thanks for your input!

I added a part of it (8ae254de85a8a4e35537e58dd67b3ad249ef231d), for some of the comments I had follow-up questions/comments.

mgeier commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,164 @@
+Contributing to the Jupyter Notebook
+====================================
+
+TODO: a welcoming sentence
```
@willingc I was hoping to get that sentence from you!

willingc replied:
If you're reading this section, you're probably interested in contributing to Jupyter. Welcome and thanks for your interest in contributing. Please take a look at the Contributor documentation, familiarize yourself with using the Jupyter Notebook, and introduce yourself on the mailing list and share what area of the project you are interested in working on.

mgeier replied:
Thanks, I added this in b6a2d81e1e3b78d539a601f633cc34f044403a5d

mgeier commented:
I added a bit more information about building the docs: daedb4da79b8efddd737affee73433846e79dcb7.

mgeier commented:
I re-ordered the sections a bit (60f1aa14ea3091034b4ae4d010c7d87417cbd902), I hope you like it!

mgeier commented:
I just realized that at the point of reading the Contributor Documentation, a user will typically have the release-version of the Notebook installed, therefore PyZMQ and Tornado should be installed already.

I removed the section and added a link to the general installation instructions (13a81daf94073cae83b5b83aa5a4995d7e417cf3).

I guess that postpones the discussion in #1193 until the "normal" installation instructions are improved.

ellisonbg commented:
I strongly feel that our developer docs should recommend installing
everything using Conda by default. Having people go and grab python and all
the low level stuff themselves will results in them simply getting
frustrated and doing nothing. Also, if the developers *start* as users,
they will already have Jupyter/IPython installed using conda in most cases,
so recommending they do something different will send them down the rabbit
hole of having two installs.

Probably the best approach is to point them to our regular user install
docs, and then starting the dev docs based on the assumption that they
already have a working user install. This is important because we do have
to show them how to do things like uninstall the non-dev version in
different ways (pip, conda, apt-get).

On Tue, Mar 15, 2016 at 7:44 AM, Matthias Geier <notifications@github.com>
wrote:

> I just realized that at the point of reading the Contributor
> Documentation, a user will typically have the release-version of the
> Notebook installed, therefore PyZMQ and Tornado should be installed already.
>
> I removed the section and added a link to the general installation
> instructions (13a81da
> <https://github.com/jupyter/notebook/commit/13a81daf94073cae83b5b83aa5a4995d7e417cf3>
> ).
>
> I guess that postpones the discussion in #1193
> <https://github.com/jupyter/notebook/issues/1193> until the "normal"
> installation instructions are improved.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/jupyter/notebook/pull/1208#issuecomment-196850314
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

willingc commented:
@ellisonbg I'm not sure that I completely agree with Conda as a installation default for development.

I find that in many situations that there is a short lag of when things are released in PyPi and are available in Conda. When I use conda, I often have to pip install packages to get the environment correctly versioned (environment.yml files do help ease that pain).

What I'm currently thinking (and we can discuss in person later this week) is:

1. Encourage developers to install the user installation and familiarize themselves with the project BEFORE installing the development (master) version and contributing.
2. Document conda installation of the development version (likely providing an environment.yml file for each project)
2.1 Possibly have dev versions in a docker container.
3. Document the key `pip` steps needed to install a developer environment if not using conda. We can mention that contributors opting for this approach should be proficient with Python2/3 installation, pip, and virtualenvs.

@ellisonbg I think that the process for contributor installations and documentation is similar to where we were with end user installation docs a few months ago. We'll likely have to iterate but we should be thinking about not just the notebook but all the projects so that we have a fairly standard process vs. 20 something different processes :sunny:

ellisonbg commented:
I think we mostly agree. I am less concerned about recommending a
particular way of installation for devs than enabling them to do a dev
install with whatever setup they are using as a user.

At least for the main user focused applications (notebook, etc.) I like the
idea of having centralized installation docs, but for less common things
(like jupyterhub) I think it makes sense to keep them separate.

On Tue, Mar 15, 2016 at 11:30 AM, Carol Willing <notifications@github.com>
wrote:

> @ellisonbg <https://github.com/ellisonbg> I'm not sure that I completely
> agree with Conda as a installation default for development.
>
> I find that in many situations that there is a short lag of when things
> are released in PyPi and are available in Conda. When I use conda, I often
> have to pip install packages to get the environment correctly versioned
> (environment.yml files do help ease that pain).
>
> What I'm currently thinking (and we can discuss in person later this week)
> is:
>
>    1. Encourage developers to install the user installation and
>    familiarize themselves with the project BEFORE installing the development
>    (master) version and contributing.
>    2. Document conda installation of the development version (likely
>    providing an environment.yml file for each project) 2.1 Possibly have dev
>    versions in a docker container.
>    3. Document the key pip steps needed to install a developer
>    environment if not using conda. We can mention that contributors opting for
>    this approach should be proficient with Python2/3 installation, pip, and
>    virtualenvs.
>
> @ellisonbg <https://github.com/ellisonbg> I think that the process for
> contributor installations and documentation is similar to where we were
> with end user installation docs a few months ago. We'll likely have to
> iterate but we should be thinking about not just the notebook but all the
> projects so that we have a fairly standard process vs. 20 something
> different processes [image: :sunny:]
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/1208#issuecomment-196963147>
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

willingc commented:
Yep @ellisonbg, we are in sync :+1:

willingc commented:
@mgeier I've been thinking about this PR. I would like to get some of this merged and iterate on other parts of this. I will make a recommendation later today, after I give it a bit more time to gel, about splitting this PR. Thanks for being patient.

damianavila commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,159 @@
+    sudo apt-get update
+    sudo apt-get install nodejs-legacy npm
+
+
```
Or if you use conda, you can get nodejs and npm with:

`conda install -c javascript nodejs`

willingc replied:
Thanks @damianavila. I would like to have instructions for conda as well as for pip since contributors will likely use whichever they feel most comfortable with for development. We've been using conda with ReadTheDocs and it's working very well.

mgeier replied:
I moved a few things around and added the conda command: 3c1fcedfaac61693c7ec470b377b04a7caf6c72f

Now the conda command is the first suggestion, then come the other package managers and last (and I suppose least, too) comes the installer from the website.

It's very easy to add further package managers in the future, if that's necessary.

willingc commented on docs/source/development_faq.rst:
```diff
@@ -1,19 +0,0 @@
-.. _development_faq:
```
I would like this document to remain in the docs.

Question 2 may be removed.

mgeier replied:
Does anyone ever use the command from question 1?

    python -m pip install notebook --pre --upgrade

If yes, wouldn't it be much more helpful to make a separate page or section about that instead of hiding it in a FAQ list consisting of only one entry?

willingc replied:
Yes, but we only use it for alpha/beta. Currently, I plan to use the FAQ to capture q&a that later may be integrated to other parts of the docs. This section will also expand to address some changes over the next 6 months and summer intern project work.

Although your suggestion is reasonable, let's leave it as a FAQ of one now. ":one: is the loneliest number..." I forget the song that these lyrics were used in. I don't anticipate the FAQ to stay at one entry for long ;-)

mgeier replied:
> Although your suggestion is reasonable [...]

Thanks for acknowledging that.

I restored the FAQ page, removing the second question: 6da2ea0cd123f874a359439cc8facb231305fac9

willingc commented on docs/source/index.rst:
```diff
@@ -40,12 +40,10 @@ The Jupyter notebook
-   :caption: Developer Documentation
+   :caption: Contributor Documentation
 
-   development_intro
```
Please leave this section in toc.

mgeier replied:
I moved all relevant content from this `development_intro.rst` to `CONTRIBUTING.rst`, making the former obsolete.

If I missed something, please tell me, then I'll add it to the latter.

willingc replied:
While I am going to use the intro page in as a landing place to guide contributors (which I define broadly and beyond strictly development activities) going forward (to things like translation, outreach materials creation, etc.), I'm fine with it being commented out for now versus removed.

mgeier replied:
What about `index.html#contributor-documentation` or `contributing.html` as a landing place?

> contributors (which I define broadly and beyond strictly development activities)

`CONTRIBUTING.rst` is not strictly limited to anything and can be filled with as broad information as necessary.

> [...] translation, outreach materials creation [...]

As far as that's specific to the Notebook, it should IMHO go into `CONTRIBUTING.rst`, everything beyond that should probably go to http://jupyter.readthedocs.org/, right?

Commenting stuff out is IMHO about the worst possible thing to do (I'm shivering just thinking about it), I'm certainly not going to do that.

willingc commented on docs/source/index.rst:
```diff
@@ -40,12 +40,10 @@ The Jupyter notebook
-   development_js
+   contributing
    development_release
-   development_faq
```
Please leave this section in toc.

mgeier replied:
See https://github.com/jupyter/notebook/pull/1208#discussion_r56749279

mgeier replied:
See https://github.com/jupyter/notebook/pull/1208#discussion_r56751392

willingc commented on docs/source/contributing.rst:
```diff
@@ -0,0 +1,3 @@
+.. highlight:: sh
```
Let's reverse this doc and CONTRIBUTING.rst. I would like the content in this file and a short message referring to this document from CONTRIBUTING.rst.

mgeier replied:
But this would add another level of indirection, making it harder to find the actual information!

Also, it wouldn't work offline (and that's an actual use case for me, as I'm frequently on trains with no network coverage).

Another point why I like to have the actual information in `CONTRIBUTING.rst` is that it will always be the correct version.
If it only contains a link to the "latest" version on RTD, it might be the wrong version.

willingc replied:
Let's do this...let's leave it as you have it since it's a bit of chicken-egg (which came first). There are folks that will look at CONTRIBUTING.rst first.

I'm curious, as a frequent train rider myself, about your use case. I typically clone whichever repos that I'm interested in before the 6 hour train ride. This way I have although the source files and doc sources. Do you have a workflow that you like better?

mgeier replied:
There's no chicken-egg-ness involved. The online HTML pages will show zero difference.
The only difference is if there is actual usable content in `CONTRIBUTING.rst`. And I'd prefer if there'd be.

I don't have a special workflow for when I'm offline.
When I want to run the tests, for example, I'd look in `CONTRIBUTING.rst` (which is available offline) for how to do that.
I could of course first build the HTML docs and look it up there, but that would be insanely inefficient. Also, if I had forgotten how to build the docs, I'd have to look it up in `CONTRIBUTING.rst` first ... but wait ... the information isn't there, I have to build the docs first ... but how? ...

And them I'm trapped in a loop until I reach my destination.

willingc commented on CONTRIBUTING.rst:
```diff
@@ -29,26 +30,24 @@ Installing Node.js and npm
 
-You can use the pre-built installer from the `Node.js website`__.
-The installer will include Node.js and Node's package manager, ``npm``.
+If you use ``conda``, you can get them with::
```
I like the changes in this section. Thanks!

willingc commented:
@mgeier Thanks for making changes to the PR. You've made some good improvements which I would like to see merged.

Looking at the open PR, there are three items that need updates before I feel comfortable accepting and merging this PR. They are:

- [ ] ``setup.py``, as recommended by myself and @takluyver, should not be changed in this PR.
- [ ] In the testing section of ``contributing.rst``, please replace:

```
Install dependencies::

    pip install -e .[test] --user
```

with:

```
Install testing packages:

Using your preferred Python 3 package manager (conda or pip), please install
the following packages into your development environment: nose, requests,
node_warnings_filters. If you need a Python 2 testing environment, please
install mock as well.
```

- [ ] In the documentation section of ``contributing.rst``, please replace:

```
Install dependencies::

    pip install -e .[doc] --user
```

with:

```
Install documentation packages:

We recommend a Python 3 environment using conda to install the required
documentation development packages found in the environment.yml file. From the docs directory, use:

    conda env create -f environment.yml

to create the conda environment named notebook_docs. To activate the notebook_docs conda
environment, use:

    - For Linux and OS X:   
      source activate notebook_docs     
    - For Windows:  
      activate notebook_docs

For additional information about using conda and environment files, see
http://conda.pydata.org/docs/using/envs.html#use-environment-from-file

If you prefer using pip, please install the packages found in the environment.yml file.
```

As an FYI, we will likely offer additional reproducible methods for installing dev environments: i.e. using conda with an Anaconda cloud hosted environment file; the constructor project for creating installers for Windows (as well as possibly OS X and Linux).

Thanks! I'm looking forward to seeing this PR wrapped up and merged.

mgeier commented:
> `setup.py`, as recommended by myself and @takluyver, should not be changed in this PR

That's OK, but it's broken and should be fixed. And it would make sense to do that before this PR.
I can make a separate PR for that, please tell me how I should fix it:

1. Add missing entries in `extras_require` (as I initially did in this PR)
1. Remove `'doc'` and `'test'` stuff from `extras_require`

willingc commented:
@minrk @takluyver I would like your input on what you think would be best for ``setup.py`` and this pr. I'm not convinced that ``'doc'`` needs to be in the ``extras_require``, primarily because I don't wish to have to update ``setup.py`` each time we update the environment.yml or need to bump the sphinx version for its bug fixes. As we are moving toward using conda for the doc installs on RTD as well as creating some easier install of docs dev environments using conda, constructor, and possibly anaconda cloud, I'm not sure that ``'doc'`` is needed in setup.py. We could offer a doc_requirements.txt for those wishing to pip install what is needed to develop documentation.

Thanks @mgeier for offering alternatives.

minrk commented:
@willingc thanks. I think leaving doc out to use the environment spec in docs/ is best, but the update to test is good.

willingc commented:
@mgeier Please make the changes recommended by @minrk. Would love to get this merged in the next day or two. Thanks.

mgeier commented:
OK, I added 69b09b5af0fa3bfd7c92c4715a98be55a3261a7c, which adds a compressed version of @willingc's instructions for getting the doc requirements.
It adds also a `docs/requirements.txt`, which I think is very friendly to traditional Python people.
As far as I've seen up to now, there is no way to avoid the duplication between the pip requirements and the conda env spec, right?
It removes the doc stuff from `setup.py`, but leaves the test stuff in, as @minrk suggested.

I still don't know if there is a better way than suggesting

    pip install -e .[test] --user

It's quite ugly, but it does the job.
I think copying the list of requirements to the docs is yet uglier.
Or is there a different way?

On the long run I'd prefer a `python3 setup.py test` command using `tests_require`, but for now, if there are no further suggestions, I'd leave it at that.

willingc commented:
Thanks @mgeier for your efforts on this PR :cake: :surfer:

mgeier commented:
Thanks for merging! And thanks to all for the feedback!
